### PR TITLE
Replace vim.fn/vim.env in vim.fs

### PR DESCRIPTION
**Pull request by mike325**
`vim.fs` functions like, `vim.fs.find`, could be useful in fast event contexts and inside uv threads,

Ex. Looking for files/dirs in slow FS or traversing deep repositories can hang the ui and can be delegated to background threads 

```lua
local function find_cpp_src(path)
    local fs = require 'vim.fs'

    local function filter(filename)
        return filename:match '.+%.cpp$' ~= nil
    end

    return vim.json.encode(fs.find(filter, { limit = math.huge, path = path }))
end

local function cb(results)
    if type(results) == type '' then
        vim.g.cpp_files = vim.json.decode(results)
    end
end

local work = vim.loop.new_work(find_cpp_src, cb)
work:queue '~/foo/'
```

Right now this is partially possible initializing `vim.env` and not using `basename/dirname` functions which include `vim.fs.find` going in upward direction

### CODE REVIEW

**Comment by lewis6991** [comment]
This will regress #17501

**Comment by gpanders** [comment]
Eliminating `vim.fn` would be good I think, for the reasons you mentioned. The concern for me is that if we roll our own implementation then we have to be very careful about catching all of the corner cases that `fnamemodify` has already learned to handle over the years. We should be able to address this with a thorough suite of unit tests, however.

**Comment by zeertzjq** [comment]
> This will regress #17501

The change is only is `fs.lua`.

**Comment by zeertzjq** [comment]
BTW existing tests haven't passed yet.

**Comment by Mike325** [comment]
> Just to note, we can now mark functions in `vim.fn` to be allowed to run in fast as long as the said functions don't alter buffer state.

Are `vim.fn`'s that don't alter the buffer's state sharable between threads too ?

**Comment by lewis6991** [comment]
Not currently.

**Comment by Mike325** [comment]
I could start writing some test to compare the outputs of `fnamemodify` basename/dirname with the new lua implementations , I don't think getting rid of `vim.env` could cause any side effect since replacing it with luv implementation should behave just as the internal `setenv()` vim function.

**Comment by gpanders** [comment]
Windows test failure is relevant.

**Comment by Mike325** [comment]
Yeah, I see, unfortunately I don't have a windows machine to test the changes, I'll try to spin up a VM to test the functions

**Comment by Mike325** [comment]
After latest changes, Windows and MacOS are still failing but I'm not sure the failures are related to my changes, I had no problems testing locally using `TEST_FILTER='vim.fs.*' make functionaltest` in windows and linux (missing macos tho)

File: runtime/lua/vim/fs.lua [hunk 1]
```diff
@@ -1,5 +1,7 @@
 local M = {}
 
+local iswin = package.config:sub(1, 1) == '\\'
```
Comment (theHamsta) [inline runtime/lua/vim/fs.lua:3]: perhaps `vim.loop.os_uname` would document itself better

**Comment by gpanders** [comment]
> After latest changes, Windows and MacOS are still failing but I'm not sure the failures are related to my changes, I had no problems testing locally using `TEST_FILTER='vim.fs.*' make functionaltest` in windows and linux (missing macos tho)

Neither macOS or Windows failures look relevant to this PR.

File: test/functional/lua/fs_spec.lua [hunk 2]
```diff
@@ -70,7 +70,7 @@ describe('vim.fs', function()
 
   describe('dirname()', function()
     it('works', function()
-      local test_windows = package.config:sub(1, 1) == '\\'
+      local test_windows = vim.loop.os_uname().sysname == 'Windows_NT'
```
Comment (gpanders) [inline test/functional/lua/fs_spec.lua:73]: In the tests we have a helper function `is_os`. Simply add

```lua
local is_os = helpers.is_os
```

to the top of the file (if it's not already there) and use it like this:

```lua
is_os('win')
```

File: runtime/lua/vim/fs.lua [hunk 3]
```diff
@@ -40,15 +42,34 @@ function M.dirname(file)
   if file == nil then
     return nil
   end
-  return vim.fn.fnamemodify(file, ':h')
+  vim.validate({ file = { file, 's' } })
+  if file:match('^%w:[\\/]?$') and iswin then
```
Comment (gpanders) [inline runtime/lua/vim/fs.lua:46]: Probably is only a minor consideration, but maybe put `iswin` in the condition first, so that it fails quickly when not on Windows (otherwise, it will run `string.match()` *every* time, even if `iswin` is false).

File: runtime/lua/vim/fs.lua [hunk 4]
```diff
@@ -40,15 +42,34 @@ function M.dirname(file)
...
+    return file:gsub('\\', '/')
+  elseif not file:match('[\\/]') then
+    return '.'
+  elseif file:match('^/[^/]+$') or file == '/' then
```
Comment (gpanders) [inline runtime/lua/vim/fs.lua:50]: Similar comment here, it's cheaper/easier to check `file == '/'` so put that in the condition first so that we don't call `string.match()` if we don't need to.

**Review by gpanders — APPROVED** [review]

File: runtime/lua/vim/fs.lua [hunk 5]
```diff
@@ -1,5 +1,7 @@
 local M = {}
 
+local iswin = vim.loop.os_uname().sysname == 'Windows_NT'
```
Comment (dundargoc) [inline runtime/lua/vim/fs.lua:3]: Don't create custom os detection please, these things tend to proliferate.

```
local iswin = is_os('win')
```

is fine.

**Review by dundargoc — DISMISSED** [review]

File: runtime/lua/vim/fs.lua (hunk 5 above)
Comment (gpanders) [reply runtime/lua/vim/fs.lua:3]: Do we have `is_os` outside of the test suite?

Comment (dundargoc) [reply runtime/lua/vim/fs.lua:3]: I can't read. I missed that this was outside of test. Ignore.

**Comment by Mike325** [comment]
@dundargoc do you mean the conditions to match the cases are difficult to read ?

**Comment by dundargoc** [comment]
@Mike325 I needed to give a reason as to why I wanted to dismiss my review, since I realized it was a mistake. I wrote that to myself as a joke. Don't worry about it, the PR is fine.
